Adds Railway deployment docs (#2927)
wasp-lang/wasp: web/docs/deployment/deployment-methods/cli.md

Author: That's an interesting idea! I'll try it out after I finish all the other comments 👍

Author: I'll jump on this after we polish the "smaller" stuff here: https://github.com/wasp-lang/wasp/pull/2957

wasp deploy fly launch my-wasp-app mia
```

> :::caution Unique Name

Author: Removing admonition since they were distracting. Content kept as it was before.

  Please do not CTRL-C or exit your terminal while the commands are running.
</small>

Two things to keep in mind:

1. Your app name (for example `my-wasp-app`) must be **unique** across all of Fly or deployment will fail.

1. If your account is a member of **more than one organization** on Fly.io, you will need to specify under which one you want to execute the command. To do that, provide an additional `--org <org-slug>` option. You can find out the names (slugs) of your organizations by running `fly orgs list`.

> Under the hood, `wasp deploy fly launch` is the same as running:

Reviewer: Actually why do we need to showcase these commands here? Maybe people get confused and run them themselves too. _Maybe they don't even change the name and region 😱_

You should include these files in your version control so that you can deploy your app with a single command in the future.

> Wasp CLI sets some [required environment variables](../../project/env-vars.md#wasp-server-env-vars) in the server app automatically, such as `DATABASE_URL`, `WASP_WEB_CLIENT_URL`, and `WASP_SERVER_URL`.

Author: I've adjusted it to flow better:

> When you run the `launch` command, Wasp CLI sets some ...

> Wasp CLI sets some [required environment variables](../../project/env-vars.md#wasp-server-env-vars) in the server app automatically, such as `DATABASE_URL`, `WASP_WEB_CLIENT_URL`, and `WASP_SERVER_URL`.

Reviewer 2: No no, I meant, the list of environment variables seems random and arbitrary to someone who doesn't know Wasp internals.

> Wasp CLI sets some [required environment variables](../../project/env-vars.md#wasp-server-env-vars) in the server app automatically, such as `DATABASE_URL`, `WASP_WEB_CLIENT_URL`, and `WASP_SERVER_URL`.

Author: Removed it 👍

> Wasp CLI sets some [required environment variables](../../project/env-vars.md#wasp-server-env-vars) in the server app automatically, such as `DATABASE_URL`, `WASP_WEB_CLIENT_URL`, and `WASP_SERVER_URL`.

Reviewer 2: I think we still don't get each other :smile: 

So, the problem is...
It's weird to say "we'll set some of the variables for you, such as..." without explicitly listing all of them and explaining why they are special. Removing the list of variables does not improve the situation, I'd argue it makes it worse.

Imagine being a user and reading this:
> When you run the launch command, Wasp CLI sets some [required environment variables](http://localhost:3002/docs/project/env-vars#wasp-server-env-vars) in the server app automatically. If your app requires any additional environment variables, use the wasp deploy fly cmd secrets set command. Read more in the [API Reference](http://localhost:3002/docs/deployment/deployment-methods/cli#flyio-cli-environment-variables).

You don't know what it sets or why it sets it. You can go to the API reference or the linked env variable page, but neither of those will answer those questions - Until you try it out, you have no idea what needs to be set and what doesn't. There's virtually no way your deployment will work from the first try.

The original version had a list of the things the command will set, that's good. The problems were:
- The "such as" part - We should list all of them and make it claer that those aren't examples and are in fact everything Wasp sets. I actually think we did list all of them but the wording wasn't clear on that.
- No explanation on why those environment variables are special.

So, I imagined something like:

> Since Wasp deployes all parts of your application and decies on their names, it also sets all environment variables needed to connect them, specifically: ...
You still have to set all other environment (e.g., ...) variables yourself

This is just a rough idea, but I trust you get what I'm going for.

  We need to do this to keep our CORS configuration up to date.
</small>

> That's it, your app should be available at `https://mycoolapp.com`! 🎉

Reviewer: I liked the 🎉, it's a easy-to-see visual signifier that you're done with the tutorial and have accomplished the task you set out to do.

:::

## Railway {#railway}

### Prerequisites

:::caution Railpack required

Wasp CLI requires that Railpack is set as the default deployment builder **for client routing to work correctly**.

Go to your [Railway account settings](https://railway.com/account/feature-flags) and enable "Default to Railpack".

> :::

Author: I've moved the Railpack warning to the prerequisites list, but I've kept the "Keep in mind" below because it's not really a prerequisite for running `wasp deploy railway`.

To deploy to Railway using Wasp CLI, make sure to:

1. Create a [Railway](https://railway.com/) account,

1. Install the [`railway` CLI](https://docs.railway.com/guides/cli#installing-the-cli) on your machine.

### Deploying

Using the Wasp CLI, you can easily deploy a new app to Railway with a single command:

```shell
wasp deploy railway launch my-wasp-app
```

<small>
  Please do not CTRL-C or exit your terminal while the commands are running.
</small>

Keep in mind that:

> 1. Your project name (for example `my-wasp-app`) must be unique across all your Railway projects or deployment will fail. This is a current limitation of the Wasp CLI and Railway integration ([#2926](https://github.com/wasp-lang/wasp/issues/2926)).

Author: We had the practice of linking to issues about things users might ask us about. For example, I'm pointing out an artificial limitation that we introduced, maybe the user wants to say "Hey make this happens pls" and that's why I'm giving them a way to do it. 

What if I put the sentence in parentheses?

> 1. Your project name (for example `my-wasp-app`) must be unique across all your Railway projects or deployment will fail. This is a current limitation of the Wasp CLI and Railway integration ([#2926](https://github.com/wasp-lang/wasp/issues/2926)).

Author: I've put in parentheses now 👍

> 1. If you are a member of multiple Railway organizations, the CLI will prompt you to select the organization under which you want to deploy your app.

Author: For me it doesn't fall under the prerequisites to use `wasp deploy railway` - it's a different kind of thing, a limitation, a tip etc. not really a blocker for using the Wasp CLI to deploy to Railway.

- `my-wasp-app-server`

Railway doesn't allow setting the database service name using the Railway CLI, so it will always be named `Postgres`.

When you run the `launch` command, Wasp CLI sets some [required environment variables](../../project/env-vars.md#wasp-server-env-vars) in the server app automatically.

If you have any additional environment variables that your app needs, read how to set them in the [API Reference](#railway-environment-variables).

### Using a Custom Domain For Your App {#railway-custom-domain}

Setting up a custom domain is a three-step process:

1. You need to add your domain to your Railway client service:

    - Go into the [Railway dashboard](https://railway.app/dashboard).
    - Select your project (for example `my-wasp-app`).
    - Click on the client service (for example `my-wasp-app-client`).
    - Go to the `Settings` tab and click `Custom Domain`.
    - Enter your domain name (for example `mycoolapp.com`) and port `8080`.
    - Click `Add Domain`.

2. You need to add the DNS records for your domain:

   _This will depend on your domain provider, but it should be a matter of adding a CNAME record with the values provided by Railway._

> 3. You need to set your new client URL as the `WASP_WEB_CLIENT_URL` environment variable (for example `https://mycoolapp.com`) for your **server service** in the Railway dashboard.

Reviewer: ```suggestion
3. To avoid [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS) errors, you need to set your new client URL as the `WASP_WEB_CLIENT_URL` environment variable (for example `https://mycoolapp.com`) for your **server service** in the Railway dashboard.
```

wasp deploy fly deploy
```

##### Environment Variables {#fly-launch-environment-variables}

###### Server

If you are deploying an app that requires any other environment variables (like social auth secrets), you can set them with the `--server-secret` option:

```
wasp deploy fly launch my-wasp-app mia --server-secret GOOGLE_CLIENT_ID=<...> --server-secret GOOGLE_CLIENT_SECRET=<...>
```

###### Client

If you've added any [client-side environment variables](../../project/env-vars#client-env-vars) to your app, make sure to pass them to the terminal session before running the `launch` command, for example:

```shell
REACT_APP_ANOTHER_VAR=somevalue wasp deploy fly launch my-wasp-app mia
```

#### `setup` command

> `setup` will create your client and server apps on Fly, and add some secrets. It does _not_ deploy the client or server apps.

Reviewer: ```suggestion
`setup` will register your client and server apps on Fly, and setup needed environment variables. It only needs to be run once, when initially creating the app. It does _not_ trigger a deploy for the client or server apps.
```

You should only run `create-db` once per app. If you run it multiple times, it will create multiple databases, but your app needs only one.
:::

#### `deploy` command

```shell
wasp deploy fly deploy
```

> The `deploy` command pushes your client and server live.

Reviewer: ```suggestion
The `deploy` command pushes your built client and server live.
```

flyctl platform regions
```

#### Multiple Fly.io Organizations

If you have multiple organizations, you can specify a `--org` option. For example:

```shell
wasp deploy fly launch my-wasp-app mia --org hive
```

#### Building Locally

Fly.io offers support for both **locally** built Docker containers and **remotely** built ones. However, for simplicity and reproducibility, the CLI defaults to the use of a remote Fly.io builder.

If you want to build locally, supply the `--build-locally` option to `wasp deploy fly launch` or `wasp deploy fly deploy`.

### Railway CLI Commands {#railway-cli}

#### `launch` command

`launch` is a convenience command that runs `setup` and `deploy` in sequence.

```shell
wasp deploy railway launch <project-name>
```

It accepts the following arguments:

> - `<project-name>` - the name of your project <Required />

Author: <img width="790" height="442" alt="SCR-20250715-lkwg" src="https://github.com/user-attachments/assets/ba898dbf-40c6-4201-b63e-023da175561f" />

Updated it to look like this 👍

It gives you the same result as running the following commands:

```shell
wasp deploy railway setup <project-name>
wasp deploy railway deploy <project-name>
```

##### Explicitly providing the Railway project ID

Wasp CLI tries to create a new Railway project named `<project-name>`, but you can use an existing project you have created on Railway by passing the `--existing-project-id [projectId]` option.

##### Environment Variables {#railway-launch-environment-variables}

> ###### Server

Reviewer: IMO Such deep nesting is reflecting the fact that we might need to split this page.

```shell
REACT_APP_ANOTHER_VAR=somevalue wasp deploy railway launch my-wasp-app
```

#### `deploy` command

The `deploy` command deploys your client and server apps to Railway.

```shell
wasp deploy railway deploy <project-name>
```

It accepts the following arguments:

- `<project-name>` - the name of your project <Required />

Run this command whenever you want to **update your deployed app** with the latest changes:

```shell
wasp deploy railway deploy my-wasp-app
```

##### Explicitly providing the Railway project ID

When you run the `deploy` command, Wasp CLI will use the Railway project that's linked to the Wasp project directory. If no Railway project is linked, the command will fail asking you to run the `setup` command first.

If you are deploying your Railway app in the CI, you can pass the `--existing-project-id [projectId]` option to tell Wasp CLI the Railway project ID to use for the deployment.

##### Other Available Options

- `--skip-client` - do not deploy the web client
- `--skip-server` - do not deploy the server

> If you've added any [client-side environment variables](../../project/env-vars#client-env-vars) to your app, make sure to pass them to the terminal session before running the `deploy` command, for example:

Author: Removed all `make sure to` to be more direct.

```shell
REACT_APP_ANOTHER_VAR=somevalue wasp deploy railway deploy my-wasp-app
```

Make sure to add your client-side environment variables every time you redeploy with the above command [to ensure they are included in the build process](../../project/env-vars#client-env-vars-1).

#### `setup` command

`setup` will create your client, server, and database services on Railway, and configure environment variables. It does _not_ deploy the client or server services.

```shell
wasp deploy railway setup <project-name>
```

It accepts the following arguments:

- `<project-name>` - the name of your project <Required />

The project name is used as a base for your server and client service names on Railway:

- `<project-name>-client`
- `<project-name>-server`

Railway will also create a PostgreSQL database service named `Postgres`.

##### Explicitly providing the Railway project ID

> You can use an existing project you have created on Railway by passing the `--existing-project-id [projectId]` option.

Author: Yes, for different commands, I want to provide the info in context at cost of being repetitive.

> You can use an existing project you have created on Railway by passing the `--existing-project-id [projectId]` option.

Reviewer 2: Fair enough!

:::caution Execute Only Once
You should only run `setup` once per app. Wasp CLI will skip creating the services if they already exist.
:::

#### Environment Variables {#railway-environment-variables}

##### Server Secrets

If your app requires any other server-side environment variables (like social auth secrets), you can set them:

1. initially in the `launch` or `setup` commands with the [`--server-secret` option](#railway-launch-environment-variables),\
   or

> 2. after the app has already been deployed go into the Railway dashboard and set them in the `Variables` tab of your server service.

Reviewer 2: This is ok, but if we want to emphasize that these are mutually exclusive, we can do something like:
> If your app requires ...., there are two ways you can set them:

Or
> If your app requires ...., you can set them either:

But this is fine too, so resolve when your happy.
